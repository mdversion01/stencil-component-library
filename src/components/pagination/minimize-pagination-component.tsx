// src/components/pagination/minimize-pagination-component.tsx
import { Component, Prop, Event, EventEmitter, h, Element } from '@stencil/core';

@Component({
  tag: 'minimize-pagination-component',
  styleUrls: ['./pagination-styles.scss'],
  shadow: false,
})
export class MinimizePagination {
  @Element() el!: HTMLElement;

  @Prop() currentPage: number = 1;
  @Prop() totalPages: number = 1;
  @Prop() goToButtons: string = '';
  @Prop() size: '' | 'sm' | 'lg' = '';
  @Prop() paginationLayout: '' | 'center' | 'end' = '';
  @Prop() plumage: boolean = false;
  // DO NOT use `id` here; use a different prop/attribute
  @Prop({ attribute: 'control-id' }) controlId?: string;

  @Event({ eventName: 'change-page' }) changePage!: EventEmitter<{ page: number }>;

  private firstPage = () => this.changePage.emit({ page: 1 });
  private prevPage = () => { if (this.currentPage > 1) this.changePage.emit({ page: this.currentPage - 1 }); };
  private nextPage = () => { if (this.currentPage < this.totalPages) this.changePage.emit({ page: this.currentPage + 1 }); };
  private lastPage = () => this.changePage.emit({ page: this.totalPages });

  render() {
    const sizeCls = this.size === 'sm' ? ' pagination-sm' : this.size === 'lg' ? ' pagination-lg' : '';
    const layoutCls = this.paginationLayout === 'center' ? ' justify-content-center'
                     : this.paginationLayout === 'end' ? ' justify-content-end' : '';
    const plumageCls = this.plumage ? ' plumage' : '';
    const controls = this.controlId ?? this.el.id; // fallback to host id

    return (
      <ul role="menubar" aria-disabled="false" aria-label="Pagination"
          class={`pagination b-pagination${sizeCls}${layoutCls}${plumageCls}`}>
        <li role="presentation" aria-hidden={this.currentPage === 1 as any}
            class={`page-item${this.currentPage === 1 ? ' disabled' : ''}`}>
          <button role="menuitem" type="button" tabIndex={0}
                  aria-label="Go to first page" aria-controls={controls}
                  class="page-link" onClick={this.firstPage}
                  disabled={this.currentPage === 1}>
            {this.goToButtons === 'text' ? 'First' : '«'}
          </button>
        </li>

        <li role="presentation" aria-hidden={this.currentPage === 1 as any}
            class={`page-item${this.currentPage === 1 ? ' disabled' : ''}`}>
          <button role="menuitem" type="button" tabIndex={0}
                  aria-label="Go to previous page" aria-controls={controls}
                  class="page-link" onClick={this.prevPage}
                  disabled={this.currentPage === 1}>
            {this.goToButtons === 'text' ? 'Prev' : '‹'}
          </button>
        </li>

        <li role="presentation" aria-hidden={this.currentPage === this.totalPages as any}
            class={`page-item${this.currentPage === this.totalPages ? ' disabled' : ''}`}>
          <button role="menuitem" type="button" tabIndex={0}
                  aria-label="Go to next page" aria-controls={controls}
                  class="page-link" onClick={this.nextPage}
                  disabled={this.currentPage === this.totalPages}>
            {this.goToButtons === 'text' ? 'Next' : '›'}
          </button>
        </li>

        <li role="presentation" aria-hidden={this.currentPage === this.totalPages as any}
            class={`page-item${this.currentPage === this.totalPages ? ' disabled' : ''}`}>
          <button role="menuitem" type="button" tabIndex={0}
                  aria-label="Go to last page" aria-controls={controls}
                  class="page-link" onClick={this.lastPage}
                  disabled={this.currentPage === this.totalPages}>
            {this.goToButtons === 'text' ? 'Last' : '»'}
          </button>
        </li>
      </ul>
    );
  }
}
