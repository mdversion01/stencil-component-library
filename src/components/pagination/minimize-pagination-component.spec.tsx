// src/components/pagination/minimize-pagination-component.spec.tsx
import { newSpecPage } from '@stencil/core/testing';
import { h } from '@stencil/core';
import { MinimizePagination } from './minimize-pagination-component';

describe('minimize-pagination-component', () => {
  it('snapshot', async () => {
    const page = await newSpecPage({
      components: [MinimizePagination],
      template: () => <minimize-pagination-component currentPage={3} totalPages={12} paginationLayout="end" size="sm" goToButtons="text" />,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('emits change-page on next click', async () => {
    const page = await newSpecPage({
      components: [MinimizePagination],
      template: () => <minimize-pagination-component currentPage={1} totalPages={2} />,
    });
    const spy = jest.fn();
    page.root!.addEventListener('change-page', (e: any) => spy(e.detail));

    const nextBtn = Array.from(page.root!.querySelectorAll('button.page-link')).find(
      b => (b.textContent || '').trim() === 'â€º' || (b.textContent || '').trim() === 'Next',
    ) as HTMLButtonElement;

    nextBtn.click();
    await page.waitForChanges();
    expect(spy).toHaveBeenCalledWith({ page: 2 });
  });
});
