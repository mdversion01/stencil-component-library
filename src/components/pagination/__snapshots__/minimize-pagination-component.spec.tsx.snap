// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minimize-pagination-component snapshot 1`] = `
<minimize-pagination-component>
  <ul aria-disabled="false" aria-label="Pagination" class="b-pagination justify-content-end pagination pagination-sm" role="menubar">
    <li class="page-item" role="presentation">
      <button aria-controls="" aria-label="Go to first page" class="page-link" role="menuitem" tabindex="0" type="button">
        First
      </button>
    </li>
    <li class="page-item" role="presentation">
      <button aria-controls="" aria-label="Go to previous page" class="page-link" role="menuitem" tabindex="0" type="button">
        Prev
      </button>
    </li>
    <li class="page-item" role="presentation">
      <button aria-controls="" aria-label="Go to next page" class="page-link" role="menuitem" tabindex="0" type="button">
        Next
      </button>
    </li>
    <li class="page-item" role="presentation">
      <button aria-controls="" aria-label="Go to last page" class="page-link" role="menuitem" tabindex="0" type="button">
        Last
      </button>
    </li>
  </ul>
</minimize-pagination-component>
`;
